internal/controller/volumereplicationgroup_controller.go

```diff
@@ -800,41 +795,267 @@ func (v *VRGInstance) separatePVCsUsingVRGStatus(pvcList *corev1.PersistentVolum
+
+	SnaphotClasses, err := v.volSyncHandler.GetVolumeSnapshotClasses()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, snapshotClass := range SnaphotClasses {
+		if v.selectVSC(snapshotClass, *storageClass) {
+			v.log.Info(fmt.Sprintf("snapshotClass Found: %s", snapshotClass.Name))
+			v.volSyncPVCs = append(v.volSyncPVCs, *pvc)
+
+			return &snapshotClass, nil
+		}
+	}
+
+	return nil, fmt.Errorf("failed to find VolumeSnapshotClass")
+}
+
+func (v *VRGInstance) selectVSC(snapshotClass snapv1.VolumeSnapshotClass,
+	storageClass storagev1.StorageClass,
+) bool {
+	snapshotClassLabels := snapshotClass.GetLabels()
+	storageClassLabels := storageClass.GetLabels()
+
+	if storageIDFromSnapshotClass, exists := snapshotClassLabels[StorageIDLabel]; exists {
+		if storageClassLabels[StorageIDLabel] == storageIDFromSnapshotClass {
+			v.log.Info(fmt.Sprintf("storageId %s from storageClass %s match found in VolumeSnapshotClass %s",
+				storageClassLabels[StorageIDLabel], storageClass.Name, snapshotClass.Name))
+
+			return true
+		}
+	}
+
+	return false
+}
+
+// nolint: cyclop
+func (v *VRGInstance) selectVRCUsingPeerClassAndSC(replicationClass volrep.VolumeReplicationClass,
+	peerClass ramendrv1alpha1.PeerClass, storageClass storagev1.StorageClass,
+) bool {
+	v.log.Info("selecting VolumeReplicationClass")
+
+	replClassLabels := replicationClass.GetLabels()
+	storageClassLabels := storageClass.GetLabels()
+
+	replicationIDMatch := false
+	storageIDMatch := false
+	storageProvisionerMatch := false
+
+	// checks if replicationIDlabel exists in the replciationclass and proceed if exists
+	if replicationID, exists := replClassLabels[VolumeReplicationIDLabel]; exists {
+		if peerClass.ReplicationID == replicationID {
+			replicationIDMatch = true
+		}
+
+		if !replicationIDMatch {
+			v.log.Info(fmt.Sprintf(
+				"replicationID %s from peerClass does not match with replicationID %s in VolumeReplicationClass %s",
+				peerClass.ReplicationID, replicationID, replicationClass.Name))
+		}
+	}
+
+	if storageIDfromReplicationClass, exists := replClassLabels[StorageIDLabel]; exists {
+		if storageIDfromReplicationClass == storageClassLabels[StorageIDLabel] {
+			storageIDMatch = true
+		}
+
+		if !storageIDMatch {
+			v.log.Info(fmt.Sprintf("storageID (%s) not found in VolumeReplicationClass (%s)", storageIDfromReplicationClass,
+				replicationClass.Name))
+		}
+	}
+
+	if storageClass.Provisioner == replicationClass.Spec.Provisioner {
+		storageProvisionerMatch = true
+	}
+
+	if !storageProvisionerMatch {
+		v.log.Info(fmt.Sprintf(
+			"provisioner %s from storageClass %s does not matches with provisioner %s from VolumeReplicationClass %s",
+			storageClass.Provisioner, storageClass.Name, replicationClass.Spec.Provisioner, replicationClass.Name))
+	}
+
+	return replicationIDMatch && storageIDMatch && storageProvisionerMatch
+}
+
+// find replication class match
+func (v *VRGInstance) findMatchingReplicationClass(peerClass *ramendrv1alpha1.PeerClass,
```

Comment:
Again, I got confused above. A function that starts with `find...` is expected to return whatever it finds, and does not modify anything.  Line 1026 is another place where we are adding PVC to `volRepPVCs`.